File: js/src/mail.js
"use strict";

const { HomeCloudError } = require("./errors");

function mailConsolePathToDataPlane(p, accountId) {
  const raw = String(p).replace(/^\/+/, "");
  const prefix = `accounts/${accountId}/mail/`;
  if (raw.startsWith(prefix)) return `/${accountId}/${raw.slice(prefix.length)}`;
  if (raw.startsWith(`${accountId}/`)) return `/${raw}`;
  throw new HomeCloudError(`Unexpected mail path for data plane: ${p}`);
}

class MailAPI {
  constructor(client) {
    this._c = client;
  }

  _useMailDataPlane() {
    return Boolean(this._c.accessKeyId && this._c.dataPlaneBases.mail);
  }

  async _mailRequest(method, consolePath, { params } = {}) {
    const accountId = this._c.accountId;
    if (this._useMailDataPlane()) {
      const dpPath = mailConsolePathToDataPlane(consolePath, accountId);
      return this._c.dataPlaneRequest("mail", method, dpPath, { params });
    }
    this._c.requireConsole();
    return this._c.consoleRequest(method, consolePath, { params });
  }

  async _mailRequestBytes(method, consolePath) {
    const accountId = this._c.accountId;
    if (this._useMailDataPlane()) {
      const dpPath = mailConsolePathToDataPlane(consolePath, accountId);
      return this._c.dataPlaneRequestBytes("mail", method, dpPath);
    }
    this._c.requireConsole();
    return this._c.consoleRequestBytes(method, consolePath);
  }

  async listMailboxes() {
    const accountId = this._c.accountId;
    const data = await this._mailRequest("GET", `accounts/${accountId}/mail/mailboxes`);
    return data.items || [];
  }

  async listMessages({
    mailboxId = null,
    folder = null,
    direction = null,
    status = null,
    search = null,
    limit = 50,
    cursor = null,
    // legacy helper used by early smoke
    mailbox = null,
  } = {}) {
    const accountId = this._c.accountId;
    if (mailbox && !mailboxId && this._useMailDataPlane()) {
      // Early MVP path: /{account}/mailboxes/{name}/messages
      const reqPath = `/${accountId}/mailboxes/${encodeURIComponent(mailbox)}/messages`;
      return this._c.dataPlaneRequest("mail", "GET", reqPath, { params: { limit } });
    }
    const params = { limit };
    if (mailboxId) params.mailbox_id = mailboxId;
    if (folder) params.folder = folder;
    if (direction) params.direction = direction;
    if (status) params.status = status;
    if (search) params.search = search;
    if (cursor) params.cursor = cursor;
    return this._mailRequest("GET", `accounts/${accountId}/mail/messages`, { params });
  }

  async getMessage(messageId) {
    const accountId = this._c.accountId;
    return this._mailRequest("GET", `accounts/${accountId}/mail/messages/${messageId}`);
  }

  async downloadAttachment(messageId, partId) {
    const accountId = this._c.accountId;
    return this._mailRequestBytes(
      "GET",
      `accounts/${accountId}/mail/messages/${messageId}/attachments/${partId}`
    );
  }
}

module.exports = { MailAPI };

File: java/src/main/java/com/homecloudlab/sdk/Mail.java
package com.homecloudlab.sdk;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Mail {
    private final HomeCloud c;

    Mail(HomeCloud c) {
        this.c = c;
    }

    private byte[] request(String method, String path, Map<String, String> q) {
        c.ensureAccountId();
        if (c.hasAccessKey() && c.hasMailDataPlaneOverride()) {
            String dp = mailConsolePathToDataPlane(path, c.accountIdOrEmpty());
            return c.dataPlaneJson("mail", method, dp, "", q, null, null);
        }
        if (c.hasAccessKey()) {
            return c.consoleSignedJson(method, path, c.accountIdOrEmpty(), q);
        }
        return c.consoleJson(method, path, true, null, q, null, null);
    }

    public List<Mailbox> listMailboxes() {
        c.ensureAccountId();
        return Json.itemsOf(request("GET", "accounts/" + c.accountIdOrEmpty() + "/mail/mailboxes", null), Mailbox.class);
    }

    public MailMessageList listMessages(ListMessagesOptions opts) {
        c.ensureAccountId();
        if (opts == null) {
            opts = ListMessagesOptions.builder().build();
        }
        int limit = opts.limit() <= 0 ? 50 : opts.limit();
        Map<String, String> q = new LinkedHashMap<>();
        q.put("limit", Integer.toString(limit));
        putIf(q, "mailbox_id", opts.mailboxId());
        putIf(q, "folder", opts.folder());
        putIf(q, "direction", opts.direction());
        putIf(q, "status", opts.status());
        putIf(q, "search", opts.search());
        putIf(q, "cursor", opts.cursor());
        byte[] raw = request("GET", "accounts/" + c.accountIdOrEmpty() + "/mail/messages", q);
        MailMessageList list = Json.decode(raw, MailMessageList.class);
        return list == null ? new MailMessageList(List.of(), "", false) : list;
    }

    public MailMessage getMessage(String messageId) {
        c.ensureAccountId();
        return Json.decode(request("GET", "accounts/" + c.accountIdOrEmpty() + "/mail/messages/" + messageId, null), MailMessage.class);
    }

    public byte[] downloadAttachment(String messageId, String partId) {
        c.ensureAccountId();
        return request("GET", "accounts/" + c.accountIdOrEmpty() + "/mail/messages/" + messageId + "/attachments/" + partId, null);
    }

    private static void putIf(Map<String, String> q, String k, String v) {
        if (v != null && !v.isEmpty()) {
            q.put(k, v);
        }
    }

    static String mailConsolePathToDataPlane(String path, String accountId) {
        String p = path;
        while (p.startsWith("/")) {
            p = p.substring(1);
        }
        String prefix = "accounts/" + accountId + "/mail/";
        if (p.startsWith(prefix)) {
            return "/" + accountId + "/" + p.substring(prefix.length());
        }
        if (p.startsWith(accountId + "/")) {
            return "/" + p;
        }
        return "/" + p;
    }
}
